### PR TITLE
Swaps event system to be push-based

### PR DESCRIPTION
**Pull request by Spoffy**

### CODE REVIEW

File: paradigm/global/functions/events/fn_event_subscribeToClients.sqf [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+    Description:
+        Causes the handler to be called, whenever specified event is received from one of the given clients.
+
+        Event can be a string, or a combination of event and topic, where topic is anything hashable.
+        - Just specifying an event means all events of that type are received
+        - Specifying an event and topic, means only events with a matching topic will cause the callback to fire
+
+        This immediately registers the handler, so that the next event received will invoke it.
+        It's safe to assume that once this function has been called, all
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_subscribeToClients.sqf]: Seems like you've forgotten to complete this.

File: paradigm/global/functions/events/fn_event_triggerGlobal.sqf [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*
+    File: fnc_event_triggerGlobal.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_triggerGlobal.sqf]: Should be fn_event_triggerGlobal.sqf

File: paradigm/global/functions/events/fn_event_triggerServer.sqf [hunk 3]
```diff
@@ -0,0 +1,33 @@
+/*
+    File: fnc_event_triggerServer.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_triggerServer.sqf]: Should be fn_event_triggerServer.sqf

File: paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf [hunk 4]
```diff
@@ -0,0 +1,36 @@
+/*
+    File: fnc_event_triggerServerAndLocal.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf]: Should be fn_event_triggerServerAndLocal.sqf

File: paradigm/global/functions/events/fn_event_triggerLocal.sqf [hunk 5]
```diff
@@ -1,12 +1,12 @@
 /*
-    File: fnc_event_trigger.sqf
+    File: fnc_event_triggerLocal.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_triggerLocal.sqf]: Should be fn_event_triggerLocal.sqf

File: paradigm/global/functions/events/fn_event_triggerTargets.sqf [hunk 6]
```diff
@@ -0,0 +1,34 @@
+/*
+    File: fnc_event_triggerTargets.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/fn_event_triggerTargets.sqf]: Should be in fn_event_triggerTargets.sqf

File: paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/*
+    File: fnc_event_remoteExec_trigger.sqf
```
Comment (cerebral-314) [inline paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf]: Should be fn_event_remoteExec_trigger.sqf

File: paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    Parameter(s):
+        _originClient - Machine ID of the client sending the event. [NUMBER]
+        _event - Triggering event. Can either be a string, or [event, topic] array, where event is a string, and topic is anything hashable. [STRING/ARRAY]
+        _data - Optional data to send to the clients. This is passed to the callbacks. [ANY]
+
+    Returns:
+        Nothing
+
+    Example(s):
+        ["myCustomEvent", 3] call para_g_fnc_event_triggerLocal;
```
Comment (cerebral-314) [inline paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf]: These examples seem to be for para_g_fnc_event_triggerLocal

**Review by cerebral-314 — COMMENTED** [review]
Looks good aside from some comment stuff.
Are we storing the events as a string array or a actual hashmap?

File: paradigm/global/functions/events/fn_event_triggerGlobal.sqf [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+params ["_event", "_data"];
+
+// Standardise event format
+if !(_event isEqualType []) then {
+    _event = [_event, ""]
+};
+
+private _hashableEvent = [_event] call para_g_fnc_event_convertEventToHashableEvent;
+
+[clientOwner, _event, _data] remoteExec ["para_g_fnc_event_remoteExec_trigger", -clientOwner];
```
Comment (veteran29) [inline paradigm/global/functions/events/fn_event_triggerGlobal.sqf]: Use `remoteExecCall` for unscheduled execution.
```suggestion
[clientOwner, _event, _data] remoteExecCall ["para_g_fnc_event_remoteExec_trigger", -clientOwner];
```

File: paradigm/global/functions/events/fn_event_triggerGlobal.sqf [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+
+[clientOwner, _hashableEvent, _event, _data] call para_g_fnc_event_callRegisteredHandlers;
```
Comment (veteran29) [inline paradigm/global/functions/events/fn_event_triggerGlobal.sqf]: Consider using `triggerLocal` to not duplicate the code.

File: paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+// Standardise event format
+if !(_event isEqualType []) then {
+    _event = [_event, ""]
+};
+
+private _hashableEvent = [_event] call para_g_fnc_event_convertEventToHashableEvent;
+
+[clientOwner, _event, _data] remoteExec ["para_g_fnc_event_remoteExec_trigger"];
+
+[clientOwner, _hashableEvent, _event, _data] call para_g_fnc_event_callRegisteredHandlers;
```
Comment (veteran29) [inline paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf]: Consider using `triggerLocal` to not duplicate the code.

File: paradigm/global/functions/events/fn_event_triggerTargets.sqf [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+params ["_event", "_data", "_targets"];
+
+// Standardise event format
+if !(_event isEqualType []) then {
+    _event = [_event, ""]
+};
+
+[clientOwner, _event, _data] remoteExec ["para_g_fnc_event_remoteExec_trigger", _targets];
```
Comment (veteran29) [inline paradigm/global/functions/events/fn_event_triggerTargets.sqf]: Use `remoteExecCall ` for unscheduled execution.
```suggestion
[clientOwner, _event, _data] remoteExecCall ["para_g_fnc_event_remoteExec_trigger", _targets];
```

File: paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+params ["_event", "_data"];
+
+// Standardise event format
+if !(_event isEqualType []) then {
+    _event = [_event, ""]
+};
+
+private _hashableEvent = [_event] call para_g_fnc_event_convertEventToHashableEvent;
+
+[clientOwner, _event, _data] remoteExec ["para_g_fnc_event_remoteExec_trigger"];
```
Comment (veteran29) [inline paradigm/global/functions/events/fn_event_triggerServerAndLocal.sqf]: Use `remoteExecCall ` for unscheduled execution.
```suggestion
[clientOwner, _event, _data] remoteExecCall ["para_g_fnc_event_remoteExec_trigger"];
```

File: paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf [hunk 14]
```diff
@@ -0,0 +1,42 @@
+/*
+    File: fnc_event_remoteExec_trigger.sqf
+    Author: Savage Game Design
+    Date: 2022-11-20
+    Last Update: 2023-04-13
+    Public: Yes
+
+    Description:
+        Remote executed from another clien to trigger the given event.
```
Comment (veteran29) [inline paradigm/global/functions/events/network/fn_event_remoteExec_trigger.sqf]: typo
```suggestion
        Remote executed from another client to trigger the given event.
```

File: paradigm/server/functions/events/internal/fn_event_registerClient.sqf [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        [_clientOwner] call para_s_fnc_event_registerClient
+ */
+
+params ["_clientMachineId"];
+
+if (isRemoteExecuted && remoteExecutedOwner isNotEqualTo 0) then {
+    _clientMachineId = remoteExecutedOwner;
+};
+
+["INFO", format ["Event system - registering client %1", _clientMachineId]] call para_g_fnc_log;
```
Comment (veteran29) [inline paradigm/server/functions/events/internal/fn_event_registerClient.sqf:28]: This function does nothing besides logging. Is it intentional?

**Review by veteran29 — CHANGES_REQUESTED** [review]

File: paradigm/server/functions/events/internal/fn_event_registerClient.sqf (hunk 15 above)
Comment (cerebral-314) [reply paradigm/server/functions/events/internal/fn_event_registerClient.sqf:28]: It appears this is modifying the parameter you're giving it to the ID of the client.

Comment (veteran29) [reply paradigm/server/functions/events/internal/fn_event_registerClient.sqf:28]: It does not as it's using `params` so the variable is "privated" to this scope.

Comment (Spoffy) [reply paradigm/server/functions/events/internal/fn_event_registerClient.sqf:28]: Yeah, it's intentionally left in. It's a handy bit of logging, and gives us something to easily hook into later.

**Review by veteran29 — APPROVED** [review]
LGTM

**Pull request merged**